src/main/java/com/example/bookstoreapp/services/impl/CartItemServiceImpl.java

```diff
@@ -12,159 +12,164 @@
 import com.example.bookstoreapp.services.UserContextService;
 import com.example.bookstoreapp.utils.IdGenerator;
 import jakarta.persistence.EntityManager;
-import jakarta.persistence.EntityTransaction;
-import jakarta.persistence.criteria.CriteriaBuilder;
-import jakarta.persistence.criteria.CriteriaDelete;
-import jakarta.persistence.criteria.CriteriaQuery;
-import jakarta.persistence.criteria.CriteriaUpdate;
-import jakarta.persistence.criteria.Predicate;
-import jakarta.persistence.criteria.Root;
+import jakarta.persistence.criteria.*;
 import lombok.extern.slf4j.Slf4j;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.stereotype.Service;
 
 import java.util.ArrayList;
 import java.util.List;
-import java.util.Map;
 
 @Service
 @Slf4j
 public class CartItemServiceImpl implements CartItemService {
 
-  @Autowired
-  private CatalogItemService catalogItemService;
+    @Autowired
+    private CatalogItemService catalogItemService;
+
+    @Autowired
+    private UserContextService userContextService;
+
+    @Autowired
+    private EntityManager entityManager;
+
+    @Autowired
+    private CartItemEntityRepository cartItemEntityRepository;
+
+    @Override
+    public CartItem addItem(Long catalogItemId) {
+        CatalogItem catalogItem = catalogItemService.findById(catalogItemId);
+        CartItemEntity savedCartItemEntity = findCartItemEntity(catalogItemId);
+        if (catalogItem.getStockQuantity() <= 0) {
+            throw new AppRuntimeException("Item is out of stock");
+        }
+        if (savedCartItemEntity == null) { // create new entity
+            CartItemEntity newCartItemEntity = new CartItemEntity();
+            newCartItemEntity.setId(IdGenerator.getLongId());
+            newCartItemEntity.setCartId(userContextService.getUserId());
+            newCartItemEntity.setQuantity(1);
+            newCartItemEntity.setCatalogItemId(catalogItemId);
+            newCartItemEntity.setState(CartItemState.BUY_NOW.getWeight());
+
+            newCartItemEntity = cartItemEntityRepository.save(newCartItemEntity);
+
+            return new CartItem().fromEntity(newCartItemEntity, catalogItem);
+        }
+
+        /* An item can have two states : One in which it has a limit & the other in which it has no limit
+        if item doesn't have any limit = In database (item_limit = 0)
+        if item has limit = In database item_limit is set to the limit count
+         */
+        // When there is no item limit
+        if (catalogItem.getItemLimit() == 0) {
```

Comment:
These if statements looks like duplicate to me. 

```java
if (catalogItem.getItemLimit() != 0) {
            if (savedCartItemEntity.getQuantity() >= catalogItem.getItemLimit()){
                throw new AppRuntimeException("You have reached the maximum limit for this item");
     }
}

savedCartItemEntity.setQuantity(savedCartItemEntity.getQuantity() + 1);
            savedCartItemEntity = cartItemEntityRepository.save(savedCartItemEntity);

```

Reply:
Applied this into the the code